File: src/tools/bin/retmt.pl
#!/usr/bin/perl -w
#
# RETrieve_Machine_Temperature, retmt:
#  Retrieve the temperatures measured within the RPI3 machine and report
#  them to Xymon.
#
# Written by W.J.M. Nelis, wim.nelis@ziggo.nl, 2017.10
#
# Modified by W.J.M. Nelis, wim.nelis@ziggo.nl, 2018.07
# - Specify the device type in the RRD file name, in this case 'cpu'.
#
use strict ;
use Time::Piece ;			# Format time

#
# Installation constants.
# -----------------------
#
my $XyDisp= $ENV{XYMSRV} ;		# Name of monitor server
my $XySend= $ENV{XYMON} ;		# Monitor interface program
my $FmtDate= "%Y.%m.%d %H:%M:%S" ;	# Default date format
   $FmtDate= $ENV{XYMONDATEFORMAT} if exists $ENV{XYMONDATEFORMAT} ;
my $HostName= `hostname` ;		# 'Source' of this test
chomp $HostName ;
my $TestName= 'env' ;			# Test name
my $ThresholdYellow= 60 ;		# Warning threshold [C]
my $ThresholdRed   = 70 ;		# Error threshold [C]

my @ColourOf= ( 'red', 'yellow', 'clear', 'green' ) ;

my $CpuFil= '/sys/class/thermal/thermal_zone0/temp' ;
my $GpuCmd= '/usr/bin/vcgencmd measure_temp' ;

#
# Global variables.
# -----------------
#
my $Now= localtime ;			# Timestamp of tests
   $Now= $Now->strftime( $FmtDate ) ;
my $Colour=  3 ;			# Test status
my $Result= '' ;			# Message to sent to Xymon
my %Temp ;				# Temperature readings

my %ErrMsg ;				# Error messages
   $ErrMsg{$_}= []		foreach ( @ColourOf ) ;

#
# Issue a message to the logfile. As this script is run periodically by Xymon,
# StdOut will be redirected to the logfile.
#
sub LogMessage {
  my $Msg= shift ;
  my @Time= (localtime())[0..5] ;
  $Time[4]++ ;  $Time[5]+= 1900 ;
  chomp $Msg ;
  printf "%4d%02d%02d %02d%02d%02d %s\n", reverse(@Time), $Msg ;
}  # of LogMessage

sub max($$) { return $_[0] > $_[1] ? $_[0] : $_[1] ; }
sub min($$) { return $_[0] < $_[1] ? $_[0] : $_[1] ; }

#
# Function InformXymon sends the message, in global variable $Result, to
# the Xymon server. Any error messages in %ErrMsg are prepended to the message
# and the status (colour) of the message is adapted accordingly.
#
sub InformXymon() {
  my $ErrMsg= '' ;
  my $Clr ;				# Colour of one sub-test

  for ( my $i= 0 ; $i < @ColourOf ; $i++ ) {
    $Clr= $ColourOf[$i] ;
    next			unless @{$ErrMsg{$Clr}} ;
    $Colour= min( $Colour, $i ) ;
    $ErrMsg.= "&$Clr $_\n"	foreach ( @{$ErrMsg{$Clr}} ) ;
  }  # of foreach
  $ErrMsg.= "\n"		if $ErrMsg ;

  $Colour= $ColourOf[$Colour] ;
  $Result= "\"status $HostName.$TestName $Colour $Now\n" .
	   "<b>Temperature sensor readings</b>\n\n" .
	   "$ErrMsg$Result\"\n" ;
  `$XySend $XyDisp $Result` ;		# Inform Xymon

  $Result= '' ;				# Reset message parameters
  $Colour=  3 ;
  $ErrMsg{$_}= []		foreach ( @ColourOf ) ;
}  # of InformXymon

#
# Function ReadSensors retrieves the temperatures and their thresholds. The
# results are stored in hash %Temp. In case of an error, the result area
# will be empty.
#
sub ReadSensors() {
  my @Lines ;				# Content of one 'file'

  %Temp= () ;				# Clear result area
  unless ( defined open(FH,'<',$CpuFil) ) {
    push @{$ErrMsg{clear}}, "Cannot read CPU temperature from $CpuFil:\n" .
			    "  $!" ;
  } else {
    @Lines= <FH> ;			# Read entire file
    unless ( @Lines == 1  and  $Lines[0] =~ m/^(\d+)/ ) {
      push @{$ErrMsg{clear}}, "Cannot read CPU temperature from $CpuFil\n:" .
			      "  Unexpected input" ;
    } else {
      $Temp{CPU}{label}= 'CPU' ;
      $Temp{CPU}{input}= sprintf( '%.1f', $1/1000 ) ;
      $Temp{CPU}{max}  = $ThresholdYellow ;
      $Temp{CPU}{crit} = $ThresholdRed ;
    }  # of else
  }  # of else

  @Lines= `$GpuCmd` ;			# Retrieve information
  if ( @Lines == 0 ) {
    push @{$ErrMsg{clear}}, "Cannot read GPU temperature from \`$GpuCmd\`:\n" .
			    "  no data returned" ;
  } else {
    chomp $Lines[0] ;
    unless ( $Lines[0] =~ m/^temp=([\d\.]+).+C$/ ) {
      push @{$ErrMsg{clear}}, "Cannot read GPU temperature from \`$GpuCmd\`:\n" .
			      "  unexpected input : $Lines[0]" ;
    } else {
      $Temp{GPU}{label}= 'GPU' ;
      $Temp{GPU}{input}= $1 ;
      $Temp{GPU}{max}  = $ThresholdYellow ;
      $Temp{GPU}{crit} = $ThresholdRed ;
    }  # of else
  }  # of else
}  # of ReadSensors

#
# Function BuildMessage formats the collected data into a nice table, performs
# the threshold checks and leaves the results in the status indicator. The
# statistics are added in the DEVMON format to be moved into an RRD.
#
sub BuildMessage() {
  my ($TempMin,$TempAvg,$TempMax)= (100,0,-100) ;	# Temperature statistics
  my $T ;				# Ref to data of one sensor
  my $Clr ;				# Status (colour) of one reading

  if ( scalar(keys %Temp) == 0 ) {
    $Result= "No data received\n" ;
    return ;
  }  # of if

 #
 # Build a table showing the various sensors, their readings and their
 # thresholds.
 #
  $Result = "<table border=1 cellpadding=5>\n" ;
  $Result.= " <tr> <th>Sensor</th> <th>Temp [C]</th> <th>Threshold [C]</th> </tr>\n" ;
  foreach ( sort keys %Temp ) {
    $T= $Temp{$_} ;			# Ref to sensor data
    $Result.= " <tr> " ;
    $Result.= "<td>$$T{label}</td> " ;	# Name of sensor

    $TempMin= min( $TempMin, $$T{input} ) ;
    $TempMax= max( $TempMax, $$T{input} ) ;
    $TempAvg+= $$T{input} ;

    $Clr= 'green' ;			# Assume temperature in range
    if      ( $$T{input} < 10 ) {	# Temperature is too low
      $Clr= 'yellow' ;
      push @{$ErrMsg{$Clr}}, "Temperature of $$T{label} is low" ;
    } elsif ( $$T{input} >= $$T{crit} ) {
      $Clr= 'red' ;
      push @{$ErrMsg{$Clr}}, "Temperature of $$T{label} is too high" ;
    } elsif ( $$T{input} >= $$T{max}  ) {
      $Clr= 'yellow' ;
      push @{$ErrMsg{$Clr}}, "Temperature of $$T{label} is high" ;
    }  # of elsif

    $Result.= "<td align='right'>$$T{input} &$Clr</td> " ;
    $Result.= "<td align='right'>$$T{max}</td> " ;
    $Result.= "</tr>\n" ;
  }  # of foreach
  $Result.= "</table>\n" ;
  $TempAvg/= scalar( keys %Temp ) ;	# Average temperature
  $TempAvg = sprintf( '%5.1f', $TempAvg ) ;

 #
 # Append the statistics, using the DEVMON format.
 #
  $Result.= "<!-- linecount=1 -->\n" ;
  $Result.= "<!--DEVMON RRD: env 0 0\n" ;
  $Result.= "DS:Temperature:GAUGE:600:-100:100 DS:MinTemp:GAUGE:600:-100:100 DS:MaxTemp:GAUGE:600:-100:100\n" ;
  $Result.= "temp.cpu $TempAvg:$TempMin:$TempMax\n" ;
  $Result.= "-->" ;
}  # of BuildMessage


#
# ----- MAIN PROGRAM -----
#
ReadSensors ;
BuildMessage ;
InformXymon ;

File: src/tools/bin/retmt.py
#!/usr/bin/env python3
# Migrated from https://wiki.xymonton.org/doku.php/monitors:retmt

import os
import re
import socket
import subprocess
from datetime import datetime

TestName = 'env'  # Test name
ThresholdYellow = 60  # Warning threshold [C]
ThresholdRed = 70  # Error threshold [C]

ColourOf = ['red', 'yellow', 'clear', 'green']

CpuFil = '/sys/class/thermal/thermal_zone0/temp'
GpuCmd = '/usr/bin/vcgencmd measure_temp'


def inform_xymon(ErrMsg):
    XyDisp = os.getenv('XYMSRV')  # Name of monitor server
    XySend = os.getenv('XYMON')  # Monitor interface program
    FmtDate = os.getenv('XYMONDATEFORMAT', '%Y.%m.%d %H:%M:%S')  # Default date format
    if not XyDisp:
        raise EnvironmentError("Environment variable XYMSRV is not set.")
    if not XySend:
        raise EnvironmentError("Environment variable XYMON is not set.")

    colour = 3  # Test status
    ErrMsg_str = ''
    for i, clr in enumerate(ColourOf):
        if ErrMsg[clr]:
            colour = min(colour, i)
            ErrMsg_str = f"{ErrMsg_str}&{clr} " + "\n&".join(ErrMsg[clr]) + "\n"

    if ErrMsg_str:
        ErrMsg_str += "\n"

    hostName = socket.gethostname().replace(".", ",")
    result = (f"status {hostName}.{TestName} {ColourOf[colour]} {datetime.now().strftime(FmtDate)}\n"
              f"<b>Temperature sensor readings</b>\n\n"
              f"{ErrMsg_str}{result}\n")

    subprocess.run([XySend, XyDisp, result], check=True)


def read_sensors():
    Temp = {}  # Temperature readings
    ErrMsg = {colour: [] for colour in ColourOf}  # Error messages

    try:
        with open(CpuFil, 'r') as f:
            lines = f.readlines()
            if len(lines) == 1 and lines[0].startswith(('0', '1', '2', '3', '4', '5', '6', '7', '8', '9')):
                temp_value = float(lines[0].strip()) / 1000
                Temp['CPU'] = {
                    'label': 'CPU',
                    'input': temp_value,
                    'max': ThresholdYellow,
                    'crit': ThresholdRed
                }
            else:
                ErrMsg['clear'].append(f"Cannot read CPU temperature from {CpuFil}:\nUnexpected input")
    except OSError as e:
        ErrMsg['clear'].append(f"Cannot read CPU temperature from {CpuFil}:\n{e}")

    try:
        lines = subprocess.check_output(GpuCmd, shell=True, universal_newlines=True).strip().splitlines()
        if len(lines) == 1 and 'temp=' in lines[0]:
            match = re.match(r'^temp=(\d+\.\d+).*C$', lines[0])
            if match:
                temp_value = float(match.group(1))
                Temp['GPU'] = {
                    'label': 'GPU',
                    'input': temp_value,
                    'max': ThresholdYellow,
                    'crit': ThresholdRed
                }
            else:
                ErrMsg['clear'].append(f"Cannot read GPU temperature from `{GpuCmd}`:\nUnexpected input : {lines[0]}")
        else:
            ErrMsg['clear'].append(f"Cannot read GPU temperature from `{GpuCmd}`:\nno data returned")
    except subprocess.CalledProcessError as e:
        ErrMsg['clear'].append(f"Cannot read GPU temperature from `{GpuCmd}`:\n{e}")
    return Temp, ErrMsg


def build_message(Temp, ErrMsg):
    if not Temp:
        return "No data received\n"

    TempMin = 100
    TempAvg = 0
    TempMax = -100

    result = ("<table border=1 cellpadding=5>\n"
             " <tr> <th>Sensor</th> <th>Temp [C]</th> <th>Threshold [C]</th> </tr>\n")

    for sensor, data in Temp.items():
        TempMin = min(TempMin, data['input'])
        TempMax = max(TempMax, data['input'])
        TempAvg += data['input']

        clr = 'green'
        if data['input'] < 10:
            clr = 'yellow'
            ErrMsg[clr].append(f"Temperature of {data['label']} is low")
        elif data['input'] >= data['crit']:
            clr = 'red'
            ErrMsg[clr].append(f"Temperature of {data['label']} is too high")
        elif data['input'] >= data['max']:
            clr = 'yellow'
            ErrMsg[clr].append(f"Temperature of {data['label']} is high")

        result = (f"{result} <tr> <td>{data['label']}</td> "
                  f"<td align='right'>{data['input']} &{clr}</td> "
                  f"<td align='right'>{data['max']}</td> "
                  "</tr>\n")

    result = f"{result}</table>\n"
    TempAvg /= len(Temp)
    TempAvg = f"{TempAvg:.1f}"

    result = (f"{result}<!-- linecount=1 -->\n"
              f"<!--DEVMON RRD: env 0 0\n"
              "DS:Temperature:GAUGE:600:-100:100 DS:MinTemp:GAUGE:600:-100:100 DS:MaxTemp:GAUGE:600:-100:100\n"
              f"temp.cpu {TempAvg}:{TempMin}:{TempMax}\n"
              "-->")
    return result


def main():
    Temp, ErrMsg = read_sensors()
    result = build_message(Temp, ErrMsg)
    inform_xymon(result, ErrMsg)


if __name__ == "__main__":
    main()
